src/ray/util/subreaper.cc

```diff
@@ -0,0 +1,177 @@
+// Copyright 2024 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/util/subreaper.h"
+
+#include <string.h>
+
+#include <vector>
+
+#ifdef __linux__
+#include <sys/prctl.h>
+#include <sys/wait.h>
+#endif
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/synchronization/mutex.h"
+#include "ray/util/logging.h"
+#include "ray/util/macros.h"
+#include "ray/util/process.h"
+
+namespace ray {
+// Platform-specific implementation of subreaper code.
+
+#ifdef __linux__
+// Linux implementation.
+
+namespace {
+
+using SignalHandlerFn = void (*)(const boost::system::error_code &, int);
+// Register a signal handler for the given signal.
+// The handler will be called with the signal_set and the error code.
+// After the handler is called, the signal will be re-registered.
+// The callback keeps a reference of the shared ptr to make sure it's not destroyed.
+void RegisterSignalHandlerLoop(std::shared_ptr<boost::asio::signal_set> signals,
+                               SignalHandlerFn handler) {
+  signals->async_wait(
+      [signals, handler](const boost::system::error_code &error, int signal_number) {
+        handler(error, signal_number);
+        RegisterSignalHandlerLoop(signals, handler);
+      });
+}
+
+void SigchldHandlerReapZombieAndRemoveKnownChildren(
+    const boost::system::error_code &error, int signal_number) {
+  if (error) {
+    RAY_LOG(WARNING) << "Error in SIGCHLD handler: " << error.message();
```

Comment:
should we just return here? what kind of error can happen?

Reply:
The only error is "signal set cancelled" as `boost::asio::error::operation_aborted` but we won't ever cancel it.

https://www.boost.org/doc/libs/1_84_0/doc/html/boost_asio/reference/basic_signal_set/async_wait.html

It does not hurt to waitpid anyway.

Reply:
Hmm I think the problem is waitpid is blocking, so if the operation is cancelled (I believe which means it can be called when a child is not dead yet), it can block raylet thread?

Reply:
can you comment in code? Also comment in the waitpid that it is non blocking (so although it is due to abort, it is okay)

Reply:
added comments for the error and WNOHANG.